### PR TITLE
Preload l0 index partitions

### PR DESCRIPTION
**Pull request by maysamyabandeh**
This fixes the existing logic for pinning l0 index partitions. The patch preloads the partitions into block cache and pin them if they belong to level 0 and pin_l0 is set.

The drawback is that it does many small IOs when preloading all the partitions into the cache is direct io is enabled. Working for a solution for that.

### CODE REVIEW

**Comment by maysamyabandeh** [comment]
The fix for small IOs will be added here: https://github.com/facebook/rocksdb/pull/2664

File: table/block_based_table_reader.cc [hunk 1]
```diff
@@ -1436,23 +1504,20 @@ BlockBasedTable::BlockEntryIteratorState::NewSecondaryIterator(
...
-      ReadLock rl(&cleaner_mu);
-      if (cleaner_set.find(offset) != cleaner_set.end()) {
-        // already have a reference to the block cache objects
-        return iter;
-      }
+  auto rep = table_->rep_;
+  if (block_map_) {
+    auto block = block_map_->find(handle.offset());
+    assert(block != block_map_->end());
+    if (LIKELY(block != block_map_->end())) {
```
Comment (siying) [inline table/block_based_table_reader.cc]: We already asserted. I don't see why we need an extra if. This if hides issues. If there is a bug here, we may never ever find it in release mode.

File: table/block_based_table_reader.cc [hunk 2]
```diff
@@ -709,9 +777,9 @@ Status BlockBasedTable::Open(const ImmutableCFOptions& ioptions,
     if (found_range_del_block && !rep->range_del_handle.IsNull()) {
       ReadOptions read_options;
       // TODO: try to use prefetched buffer too.
-      s = MaybeLoadDataBlockToCache(rep, read_options, rep->range_del_handle,
-                                    Slice() /* compression_dict */,
-                                    &rep->range_del_entry);
+      s = MaybeLoadDataBlockToCache(
+          nullptr /*prefetch_buffer*/, rep, read_options, rep->range_del_handle,
```
Comment (siying) [inline table/block_based_table_reader.cc]: `prefetch_buffer.get()` and removed the TODO in the previous line.

File: table/block_based_table_reader.cc [hunk 3]
```diff
@@ -740,21 +808,22 @@ Status BlockBasedTable::Open(const ImmutableCFOptions& ioptions,
...
-          level == 0) {
-        index_entry = &rep->index_entry;
-      }
+      CachableEntry<IndexReader> index_entry;
       unique_ptr<InternalIterator> iter(
-          new_table->NewIndexIterator(ReadOptions(), nullptr, index_entry));
+          new_table->NewIndexIterator(ReadOptions(), nullptr, &index_entry));
+      index_entry.value->CacheDependencies(pin);
+      if (pin) {
+        rep->index_entry = index_entry;
```
Comment (siying) [inline table/block_based_table_reader.cc]: nit, `rep->index_entry = std::move(index_entry)`?

File: table/block_based_table_reader.cc [hunk 4]
```diff
@@ -740,21 +808,22 @@ Status BlockBasedTable::Open(const ImmutableCFOptions& ioptions,
...
-      CachableEntry<IndexReader>* index_entry = nullptr;
-      if (rep->table_options.pin_l0_filter_and_index_blocks_in_cache &&
-          level == 0) {
-        index_entry = &rep->index_entry;
-      }
+      CachableEntry<IndexReader> index_entry;
       unique_ptr<InternalIterator> iter(
-          new_table->NewIndexIterator(ReadOptions(), nullptr, index_entry));
+          new_table->NewIndexIterator(ReadOptions(), nullptr, &index_entry));
+      index_entry.value->CacheDependencies(pin);
```
Comment (siying) [inline table/block_based_table_reader.cc]: Is it called for both of partitioned index case and non-partitioned case? If it can be called in non-partitioned case, does it prefetch all the data blocks instead?

File: table/block_based_table_reader.cc (hunk 1 above)
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: assert works only in debug mode. In release mode, if the condition is not valid, we do not return in this branch and the ROCKS_LOG_ERROR will be called.

File: table/block_based_table_reader.cc (hunk 4 above)
Comment (siying) [inline table/block_based_table_reader.cc]: Got it.

File: table/block_based_table_reader.cc (hunk 1 above)
Comment (siying) [inline table/block_based_table_reader.cc]: yes, but nobody is actively looking at logs for failures. This is not a place with visibility. If you don't feel comfortable with move it, I suggest we call abort() or something like that here.

Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: But a bug here would only cause performance issues and does not qualify taking down the service for it. I think we should understand that at each point of time there are many bugs in our code base and we should provide mechanisms to catch them but we do not have to crash the system each time one shows up.
About the existing approach here, I believe we have monitoring that aggregates the errors reflected on logs. Whether we use that often enough is one thing but using logs to record some unexpected behaviors can still give visibility for non-fatal errors.

Comment (siying) [inline table/block_based_table_reader.cc]: @maysamyabandeh please note in this code base, we don't do that. There are so many asserts in this code base and we don't try to work around when those asserts don't satisfy. I understand that your confident level of your new code is lower than existing code. That's why I suggest a fail fast solution. If your confident to the code quality is so low that you are concerning the fail fast happens so often that can cause trouble to users, I hope you add more tests to bring up your confident level, including adding it to the daily stress test coverage.

I suggest we invest on more robustness of code to handle hardware failures and user errors, rather than protecting from our own bugs.

Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: I still believe bugs in the code are a never-ending reality and software should be able to survive them. Anyway, I go ahead with your initial suggestion of removing the if statement.

File: HISTORY.md [hunk 5]
```diff
@@ -1,5 +1,7 @@
 # Rocksdb Change Log
 ## Unreleased
+
+## 5.7.8 (08/14/2017)
```
Comment (sagar0) [inline HISTORY.md]: 5.7.8 ?

Comment (maysamyabandeh) [inline HISTORY.md]: 5.8 is not landed yet: https://github.com/facebook/rocksdb/pull/2738

File: include/rocksdb/version.h [hunk 6]
```diff
@@ -5,7 +5,7 @@
 #pragma once
 
 #define ROCKSDB_MAJOR 5
-#define ROCKSDB_MINOR 7
+#define ROCKSDB_MINOR 8
```
Comment (maysamyabandeh) [inline include/rocksdb/version.h]: apparently the revert is somehow lost. I send another PR to fix this.

File: HISTORY.md (hunk 5 above)
Comment (maysamyabandeh) [inline HISTORY.md]: ah: https://github.com/facebook/rocksdb/pull/2661/files#r134051591

Comment (sagar0) [inline HISTORY.md]: Ah, seems like it was a mixup that got checked in. And hence the confusion.
